network/p2p/network.go

```diff
@@ -0,0 +1,223 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/metric"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/version"
+)
+
+var (
+	_ validators.Connector = (*Network)(nil)
+	_ common.AppHandler    = (*Network)(nil)
+	_ NodeSampler          = (*Network)(nil)
+)
+
+// ClientOption configures Client
+type ClientOption interface {
+	apply(options *ClientOptions)
+}
+
+type clientOptionFunc func(options *ClientOptions)
+
+func (o clientOptionFunc) apply(options *ClientOptions) {
+	o(options)
+}
+
+// WithNodeSampler configures the sampling strategy for Client.AppRequestAny
+func WithNodeSampler(nodeSampler NodeSampler) ClientOption {
+	return clientOptionFunc(func(options *ClientOptions) {
+		options.NodeSampler = nodeSampler
+	})
+}
+
+// ClientOptions holds client-configurable values
+type ClientOptions struct {
+	// NodeSampler is used to select nodes to route Client.AppRequestAny to
+	NodeSampler NodeSampler
+}
+
+// NewNetwork returns an instance of Network
+func NewNetwork(
+	log logging.Logger,
+	sender common.AppSender,
+	metrics prometheus.Registerer,
+	namespace string,
+) *Network {
+	return &Network{
+		log:       log,
+		sender:    sender,
+		metrics:   metrics,
+		namespace: namespace,
+		router:    newRouter(log),
+		peers:     set.SampleableSet[ids.NodeID]{},
+	}
+}
+
+// Network maintains state of the peer-to-peer network and any in-flight
+// requests
+type Network struct {
+	log       logging.Logger
+	sender    common.AppSender
+	metrics   prometheus.Registerer
+	namespace string
+
+	*router
+
+	lock  sync.RWMutex
+	peers set.SampleableSet[ids.NodeID]
+}
+
+func (n *Network) Connected(_ context.Context, nodeID ids.NodeID, _ *version.Application) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	n.peers.Add(nodeID)
+	return nil
+}
+
+func (n *Network) Disconnected(_ context.Context, nodeID ids.NodeID) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	n.peers.Remove(nodeID)
+	return nil
+}
+
+// Sample returns a pseudo-random sample of up to limit peers
+func (n *Network) Sample(_ context.Context, limit int) []ids.NodeID {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+
+	return n.peers.Sample(limit)
+}
+
+// RegisterAppProtocol reserves an identifier for an application protocol and
+// returns a Client that can be used to send messages for the corresponding
+// protocol.
+func (n *Network) RegisterAppProtocol(handlerID uint64, handler Handler, options ...ClientOption) (*Client, error) {
+	// TODO refactor router
```

Comment:
I think previously it was acceptable since there were only two abstractions of router + client, but now that we're introducing a third abstraction of a generic networking layer I'm not in love with how the router lock is being touched by other structs